### PR TITLE
feat: Extract design tokens to JSON

### PR DESCRIPTION
**Pull request by aaronccasanova**
### WHAT is this pull request doing?

This PR extracts the Polaris design tokens, previously implemented in TypeScript, to JSON files.
- Converted TypeScript design tokens to JSON
- Created JSON Schema's for all token groups
- Added `ajv-cli` to validate JSON token files
- Added a `validate-tokens` script and integrated in our CI

#### Follow up tasks
- Update Polaris `CustomProperties` component to use these JSON tokens
- Update Polaris `custom-properties-allowed-list` Stylelint plugin to use these JSON tokens

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
- [ ] For visual design changes, ping @ sarahill to update the Polaris UI kit

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -66,7 +66,8 @@
     "readme-update-version": "node ./scripts/readme-update-version",
     "version": "yarn run readme-update-version",
     "storybook": "start-storybook -p 6006 --quiet",
-    "storybook:build": "yarn run copy-polaris-tokens && build-storybook -o build-internal/storybook/static"
+    "storybook:build": "yarn run copy-polaris-tokens && build-storybook -o build-internal/storybook/static",
+    "validate-tokens": "ajv validate -s tokens/schema.token-group.json -d 'tokens/_!(color)*.json' && ajv validate -s tokens/schema.color.json -d 'tokens/_color.json'"
```
Comment (alex-page) [inline package.json]: Could this be simpler by breaking the color files into individual JSON files? `color.light.json`

**Review by alex-page — APPROVED** [review]
Love this. One thought, we separated the `.ts` files by adding the `_`. Now it's not as needed as the tokens are in `.json` files. We could just remove the `_` now.

File: package.json (hunk 1 above)
Comment (aaronccasanova) [reply package.json]: Great idea 👍

Comment (aaronccasanova) [reply package.json]: This will allow us to remove the `color` specific json schema 🙌

Comment (aaronccasanova) [reply package.json]: Okay [this commit](https://github.com/Shopify/polaris-react/pull/4809/commits/32b3ad65e766fcd16523f3f76950c49ba20ebef0):
- Removed the `_` prefix from all token files
- Extracted color scheme tokens into individual files (e.g. `color.light.json` & `color.dark.json`)
- Simplified the `validate-tokens` script

File: tokens/schemas/token-group.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "title": "Token group schema",
+  "type": "object",
+  "additionalProperties": {
+    "type": "string"
+  }
+}
```
Comment (aaronccasanova) [inline tokens/schemas/token-group.json]: If folks haven't worked with JSON Schema this is validating that our tokens are objects with a depth of 1 whose values are strings. Here is an example of the error thrown if we violate this constraint:
Screen Shot 2021-12-08 at 3 13 41 PM

**Comment by BPScott** [comment]
I'd love to see a demo of how these files can get included by the source code

**Comment by aaronccasanova** [comment]
> I'd love to see a demo of how these files can get included by the source code

@BPScott Your in luck! I just submitted a PR to update the `CustomProperty` component: https://github.com/Shopify/polaris-react/pull/4811

**Review by kyledurand — COMMENTED** [review]
This looks good but we're back to px values for spacing and shape. Can we follow up to be sure we're converting these to rems?

**Comment by alex-page** [comment]
> This looks good but we're back to px values for spacing and shape. Can we follow up to be sure we're converting these to rems?

If possible I would prefer the values to be pixel based in JSON file and converted to rems in ts. Unsure how complex that would be though. Happy to ship rem values in JSON to speed this up for now.

**Comment by kyledurand** [comment]
> If possible I would prefer the values to be pixel based in JSON file and converted to rems in ts. Unsure how complex that would be though. Happy to ship rem values in JSON to speed this up for now.

Pretty much how our convo went in slack. Namely because rems are pretty hard to read, hard to convert from figma and that could lead to a misunderstanding of the values and increased friction in use / contribution. 

I can't see it being too difficult to build but getting it to be performant could get tricky

**Comment by aaronccasanova** [comment]
@kyledurand @alex-page Here is my first pass: https://github.com/Shopify/polaris-react/pull/4811/commits/985631405e3cc8f550ccb68d411b589a3bb52bb8. Thoughts?

File: .vscode/settings.json [hunk 3]
```diff
@@ -23,5 +23,6 @@
     "{build,build-internal}/": true,
     "examples/*/build": true
   },
-  "typescript.tsdk": "./node_modules/typescript/lib"
+  "typescript.tsdk": "./node_modules/typescript/lib",
+  "cSpell.words": ["semibold"]
```
Comment (aaronccasanova) [inline .vscode/settings.json]:
```suggestion
  "typescript.tsdk": "./node_modules/typescript/lib"
```

**Comment by aaronccasanova** [comment]
In an unfortunate git accident this PR was wiped and automatically closed by GitHub. Luckily all the commits we're preserved in the [Update CustomProperty component PR# 4811 ](https://github.com/Shopify/polaris-react/pull/4811). There doesn't appear to be a way to restore this PR and thus the contents have been moved to the following PR: https://github.com/Shopify/polaris-react/pull/4818

**Pull request closed without merging**